apollo-federation/src/query_plan/fetch_dependency_graph.rs

```diff
@@ -1158,6 +1286,339 @@ impl FetchDependencyGraph {
+        node_id: NodeIndex,
+        child_id: NodeIndex,
+    ) -> Result<bool, FederationError> {
+        let node = self.node_weight(node_id)?;
+        let child = self.node_weight(child_id)?;
+        let parent_relation = self.parent_relation(child_id, node_id);
+
+        Ok(node.subgraph_name == child.subgraph_name
+            && node.defer_ref == child.defer_ref
+            && parent_relation.is_some_and(|r| r.path_in_parent.is_some()))
+    }
+
+    /// We only allow merging sibling on the same subgraph, same "merge_at" and when the common parent is their only parent:
+    /// - there is no reason merging siblings of different subgraphs could ever make sense.
+    /// - same "merge_at" paths ensures that we can merge the inputs and selections without having to worry about those
+    ///   not being at the same level (hence the empty path in the call to `merge_in_internal` below). In theory, we could
+    ///   relax this when we have the "path in parent" for both sibling, and if `sibling_to_merge` is "deeper" than `this`,
+    ///   we could still merge it in using the appropriate path. We don't use this yet, but if this get in the way of
+    ///   some query plan optimisation, we may have to do so.
+    /// - only handling a single parent could be expanded on later, but we don't need it yet so we focus on the simpler case.
+    fn can_merge_sibling_in(
+        &self,
+        node_id: NodeIndex,
+        sibling_id: NodeIndex,
+    ) -> Result<bool, FederationError> {
+        let node = self.node_weight(node_id)?;
+        let own_parents: Vec<ParentRelation> = self.parents_relations_of(node_id).collect();
+
+        let sibling = self.node_weight(sibling_id)?;
+        let sibling_parents: Vec<ParentRelation> = self.parents_relations_of(sibling_id).collect();
+        Ok(node.defer_ref == sibling.defer_ref
+            && node.subgraph_name == sibling.subgraph_name
+            && node.merge_at == sibling.merge_at
+            && own_parents.len() == 1
+            && sibling_parents.len() == 1
+            && own_parents[0].parent_node_id == sibling_parents[0].parent_node_id)
+    }
+
+    fn can_merge_grand_child_in(
+        &self,
+        node_id: NodeIndex,
+        grand_child_id: NodeIndex,
+    ) -> Result<bool, FederationError> {
+        let grand_child_parent_relations: Vec<ParentRelation> =
+            self.parents_relations_of(grand_child_id).collect();
+        if grand_child_parent_relations.len() != 1 {
+            return Ok(false);
+        }
+
+        let node = self.node_weight(node_id)?;
+        let grand_child = self.node_weight(grand_child_id)?;
+        let grand_child_parent_parent_relation =
+            self.parent_relation(grand_child_parent_relations[0].parent_node_id, node_id);
+
+        let (Some(node_inputs), Some(grand_child_inputs)) = (&node.inputs, &grand_child.inputs)
+        else {
+            return Ok(false);
+        };
+
+        Ok(node.subgraph_name == grand_child.subgraph_name
```

Comment:
yeah, just double checked and it looks like there is just a single call to this method which explicitly checks for the matching subgraph names on the if condition. Unsure why it was there instead of moving it to the target method.